content/blog/nats-operator-and-service-accounts.md

```diff
@@ -0,0 +1,337 @@
+Listening on [foo.bar]
+```
+
+Since config reload was enabled, at any time we can update the
+authorization configuration for the cluster and modify the
+permissions.  For example, if the `NatsServiceRole` is deleted while a
+client is connected, then the client will be disconnected from the
+cluster since the account is no longer valid:
+
+```sh
+$ kubectl -n nats-io delete natsservicerole nats-user
+natsservicerole.nats.io "nats-user" deleted
+
+# (in the attached container)
+/go # nats-sub -s nats://nats-user:`cat /var/run/secrets/nats.io/token`@example-nats:4222 hello.world
+Listening on [hello.world]
+Got disconnected!
+```
+
+Similarly, we can modify the permissions in the `NatsServiceRole` and the servers will update
+the permissions and reload. Config reload events can be confirmed from the logs from a NATS server.
+
+```sh
+kubectl -n nats-io logs example-nats-1 nats
+...
+[6] 2018/07/19 01:05:43.469246 [INF] Reloaded server configuration
+```
+
+### How it works
+
+As previously mentioned, the current implementation of the integration
+relies on the `TokenRequest` and `PodShareProcessNamespace` alpha features
+that were made available in the Kubernetes v1.10 release.
+
+The [TokenRequest API](https://github.com/kubernetes/community/blob/4a41674f6fb6fa69b3dfaf15267d908e909685d7/contributors/design-proposals/auth/bound-service-account-tokens.md) feature which allows issuing new tokens related
+to a `ServiceAccount` but with a different audience as the one used for
+the API server.  _Customizing the audience_ is very important in this
+case, since using the bare original token from `ServiceAccount` can
+make us run into security issues since this means that any container
+with the token would have the same permissions to get access to the
+Kubernetes API.
+
+```go
+// Issue token with audience set for the NATS cluster in this namespace only,
+// this will prevent the token from being usable against the API Server.
+ar := &authenticationv1.TokenRequest{
+	Spec: authenticationv1.TokenRequestSpec{
+		Audiences: []string{fmt.Sprintf("nats://%s.%s.svc", clusterName, ns)},
+        // ...
+	},
+}
+tr, err := kubecli.ServiceAccounts(ns).CreateToken(sa.Name, ar)
+if err != nil {
+	return err
+}
+```
+
+This token is then stored in a secret which can be mounted by a Pod
+(relatedly, there is a new feature in Kubernetes v1.11 called [Service
+Account Volume Projections](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-volume-projection) which would make it possible to not having
+to rely on a secret this way but more on that below).  Then the
+configuration of the NATS Server is updated including the newly mapped
+service account + nats service role.
+
+To be able to update the configuration of a cluster with zero
+downtime, assuming that the [PodShareProcessNamespace](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/) feature flag has
```

Comment:
To be able to update the configuration of a cluster with zero
+downtime, assuming the [PodShareProcessNamespace](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/)